### ISSUE TITLE
Enhancements and Suggestions Mega-Thread - Leave your ideas here, instead of making an issue!

### ISSUE THREAD

**Issue by TheWizardLizard** [sstixrud/WeaponCore#187]
# Welcome to the Enhancements and Suggestions Mega-Thread 

This issue serves to host all the past, and future suggested additions to WeaponCore

If you feel like you have a cool idea, leave it in the comments! It might just get added to the list below!

## User Interface / Experience

- [ ] [Reloading Time Left Indicator](https://github.com/sstixrud/WeaponCore/issues/176)
( More options for the current loading indicator weapons have )

- [ ] [Grid Type Targeting Options](https://github.com/sstixrud/WeaponCore/issues/173)
( Re-implement keens grid size targeting in WC )

- [ ] [Solid Background for MMB menu](https://github.com/sstixrud/WeaponCore/issues/119)
( This will be checked once menu rewrite is done ) 

- [ ] [Sound played on Ammo switch, and visual indicator for what ammo is selected](https://github.com/sstixrud/WeaponCore/issues/115)

- [ ] [Weapon Sequencing / Salvoing](https://github.com/sstixrud/WeaponCore/issues/181)

## Configuration Additions (for modders)

- [ ] [Subpart animation based on inventory fill level / ammo in magazine](https://github.com/sstixrud/WeaponCore/issues/164)

- [ ] [LockOnFocus should not automatically fire on targets](https://github.com/sstixrud/WeaponCore/issues/150)

- [ ] [Voxel damage extension to API](https://github.com/sstixrud/WeaponCore/issues/26)

- [ ] [Decals on impacted grids](https://github.com/sstixrud/WeaponCore/issues/124)

- [ ] [Noise played/looped on target lock/focus](https://github.com/sstixrud/WeaponCore/issues/170)

- [ ] [Jump-drive explosion mod shrapnelling explosion type](https://github.com/sstixrud/WeaponCore/issues/177)

- [ ] Upgrade Modules

- [ ] Mine Persistence

- [ ] EWAR scrambling projectiles

- [ ] Turret camera solution (for visual indicators)

## Programmable Block API

- [ ] [Orientation Matrix](https://github.com/sstixrud/WeaponCore/issues/182)

**Comment by OwendB1** [comment]
Got one, how about the ability for modders to set an aim cone in front of the weapon so it can only lock onto things that are right in front of the weapon inside the aim cone. Maybe with like a soft detach meaning that you would lose lock after your target hasnt been inside the aimcone for x amounts of seconds.

**Comment by TheWizardLizard** [comment]
added

**Comment by TheWizardLizard** [comment]
Adding my own enhancement idea. Being able to slave a gimbal / turreted weapon(s) to the center of a camera while its being used would be cool. 

Context: https://youtu.be/9COry2Thsqw

**Comment by Qarannia** [comment]
The ability to target and fire on friendly targets. Apart from the sheer entertainment factor I find myself constantly having to create ships when doing weapons testing. Setting them to SPRT ownership each time makes it harder to adjust things later without cutting and pasting a new ship in. But the real value for me would be in a MP world so friends on the same faction could duel.

**Comment by BlueDevil** [comment]
Would love to be able to rebind more keys. Having the reticle pop up ever time I go downwards in my ships due to my own binds is a bit distracting at times.

**Comment by NullMagic** [comment]
So I've been playing with the Defense Shields mod, which I love as it gives me a layer of protection that will allow me time to escape without sacrificing something that might be necessary/resource intensive. However I've come to see that there are two ends of the scale for the mod, there's the small ship scale (call it scout class or smaller) which has very little shield strength, and then you have Planetary Base shields which (assuming you can somehow pump out enough power) can become so strong that you need dozens of weapons to put a dent in them.

Now, the problem I'm seeing is that weapons have a set amount of damage that they always do, a laser on a Scout Class ship is going to always do the same amount of damage as a laser on a Capital Class ship, which means that while two scout ships might be able to blow through each other's shields in moments (because the weapons can do an amount of damage disproportionate to the ship size) they can't even put a dent in the capital class ship's shields.

Now, this doesn't seem like much of a problem, obviously a scout shouldn't outperform a capital class ship. But what happens when it's two capital class ships going at it? Or a Capital Class ship against a Space or Ground Station? Unless you have progressively bigger weapons you end up studding the large ship/base with _dozens_ of the same weapons over and over again, and that's where the problems start coming in.

In order to bring down the shield of a powerful ship, you need equally powerful weapons, either in quality or in quantity, and when quantity is your only (or best) option things can get bad, especially when you're looking at weapons with intense and bright vfx, such as many beam weapons. I've personally watched my game slow to a crawl when multiple ships have engaged in an all out slug fest in the nearby area.

So, I asked myself, how do I fix this conundrum? If I increase the damage outputs (but don't increase the costs in energy, or materials) then small ship shields become essentially worthless, because their shields simply aren't powerful enough to keep up with the high damage output that would be required for two large ships to have a battle (without overtaxing the game). And if I don't then I have to create entirely new versions of the same weapon with larger requirements and higher damage.

 (It should be noted that I am not opposed to the creation of more powerful and expensive weapons. However, I very much enjoy playing with a mod that randomizes the weapons on NPC ships, allowing even the vanilla ships to use modded weapons. If I create, or download, more powerful weapons I can easily end up facing the same issue as if I simply increased the damage output on the existing weapon. I.e. a scout ship with the armament of a capital ship, making it absurdly powerful for its size [because it's a capital class weapon] or completely worthless and unchallenging [because it doesn't have the resources to fire said weapon]).

So, I had a think about the problem, and realized that the Defense Shields mod already has the solution, it simply needs to be applied to weapons as well.

My suggestion is thus: A way to scale damage output based on the power generation of the ship, the same way that Defense Shields scale their shield strength based on power generation.

With this change you'd be able to create a weapon that scales easily to the size of the ship, letting your Capital Class destroyers have the firepower they need with only a few weapons, while also maintaining the balance that should exist between two small ships fighting each other (Or small ships trying to punch outside their weight limit). In addition, more weapons on the ship will force the power generation to be split more ways, making it a losing proposition to increase your weapon amount in the hope of increasing your damage output, at which point the only reason to have several weapons is for Point Defense purposes. (Also, it creates the interesting dynamic of having to decide where to send the majority of your power generation. Full power to shields? Re-rout power to weapons? Or do you strike a balance between the two? You already have the option to determine how much power generation to devote to shield regen, and there's very little reason not to make it as high as you safely can.)

Anyway, it's just an idea I had earlier, I don't even know if it's possible, but I felt it was an idea I needed to share. I also hope that wasn't too long winded, I know I can get a bit wordy (sorry). I also recognize that this change would be most useful for Energy Weapon mods, and may leaving more conventional weapons hanging (though arguments could possibly be made for rail guns).

**Comment by Qarannia** [comment]
Since 1.5(7), or the replacement of the MMB menu, not sure which, I have found an odd behaviour in my WC mod that uses multiple ammo types. Everything works normally until you start cycling ammo. After cycling ammo, the next time you use "click to fire" it will fire the entire magazine at once instead of one at a time.

The other thing I've found is that when I pull weapons to the hotbar one by one they have the "Shoot Once" option on the menu, but if the weapons are grouped, that option is no longer available.

**Comment by TheWizardLizard** [comment]
I think that a weapon should keep its "Click to fire" state on when they leave a cockpit, and should only **_temporarily_** disable if a user is in the escape menu or the ship menu

**Comment by dbenson24** [comment]
Would like the ability to have heat from a gun spread to neighboring guns when fired to provide a more natural way of discouraging players from building walls of guns

**Comment by UndefinedDeluxe** [comment]
Vanilla-like usage:
On menu opening, and a certain tab opened as well; create an automatic group to be dragged to or double clicked onto the hotbar for firing a specific type of weapon
weapons should not automatically be able to disable their own "click to fire" state when being used (AWG tank weapon WC pack, firing and then suddenly unable to, already reloaded, toolbar says "off" and no indication aside from that)

**Comment by mmmaxwwwell** [comment]
When the mod is updated before the server is rebooted, there is a red message shown to players saying version mismatch. This should be sent to the server logs or handled some other way than notifying the end users.

**Comment by MajorXAcE** [comment]
Would like the ability for reload time to be modular or linked to damage. Like for weapons with long cooldowns, you could use a slider(maybe linked to damagemodifier) to have faster reloads at the expense of damage. Or at least some way for the ammo to modify the reload time.

**Comment by rcdailey** [comment]
Why is grid type target option checked? I don't see those options yet.

**Comment by 0x00002a** [comment]
Would be great to have a way to explicitly ***not*** target a grid. When boarding a ship in survival (MES makes this a very common occurance) I have to turn all my weapons off because otherwise they willl rip apart my salvage. This leaves me a sitting duck for enemies and prevents me bringing stuff back to my main base (unless I turn those turrets off too which, with Corruption, I really don't wanna do). A way to broadcast the blacklisted state would be nice too, since otherwise there would still be the problem of transporting the disabled ship through areas covered by other turrets (although this might be doable with a script?).

Also I cannot find the grid type targeting options either.

**Comment by FirebladeProduction** [comment]
Simple but welcomed
weapon core functionality for handheld weapons with the addition to define suit power and suit gas usage for those handheld weapons as an alternativ/addition to physical ammo use.
examples for what we could use it for:
Handheld homing rocketlauncher
survival energy pistol
Hande´held Remote Targeter linked to a guided weapon in radio range

**Comment by sed-and-awk** [comment]
More api options with jakira’s water mod. Would love being able to make torpedos and such
